### PR TITLE
ISPN-5523 Improvements are Netty byte buffer usages

### PR DESCRIPTION
**Pull request by galderz**
The issue itself we have not been able to fix it since it appears to be a limitation of the current eager near cache architecture. Some fixes have been added which will benefit the code base in general.

### CODE REVIEW

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala [hunk 1]
```diff
@@ -23,19 +22,39 @@ object KeyValueVersionConverterFactory {
...
+   private def putLong(l: Long, offset: Int, out: Array[Byte]): Int = {
+      out.update(offset, (l >> 56).toByte)
+      out.update(offset + 1, (l >> 48).toByte)
+      out.update(offset + 2, (l >> 40).toByte)
+      out.update(offset + 3, (l >> 32).toByte)
+      out.update(offset + 4, (l >> 24).toByte)
+      out.update(offset + 5, (l >> 16).toByte)
+      out.update(offset + 6, (l >> 8).toByte)
+      out.update(offset + 7, l.toByte)
+      offset + 8
```
Comment (wburns) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala]: This line is redundant did you mean to return it?

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+   static int NUM_CLIENTS = 3;
+   static int NUM_THREADS_PER_CLIENT = 10;
+   static AtomicInteger ID = new AtomicInteger();
+   static ExecutorService EXEC = Executors.newCachedThreadPool();
+
+   static final int NUM_OPERATIONS = 10_000_000;
+   static final int NUM_KEYS_PRELOAD = 1_000;
+   static final int KEY_RANGE = 1_000;
+   static final Random R = new Random();
```
Comment (wburns) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java]: ThreadLocalRandom would be better here

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+   @AfterClass
+   public static void shutdownExecutor() {
+      EXEC.shutdown();
+   }
+
+   EmbeddedCacheManager createCacheManager() {
+      return TestCacheManagerFactory.createCacheManager(hotRodCacheConfiguration());
+   }
+
+//   HotRodServer createHotRodServer(EmbeddedCacheManager cm) {
```
Comment (wburns) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java]: remove?

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            remote.putAll(map);
+         }
+      });
+   }
+
+   static abstract class Runner implements Callable<Void> {
+
+      final CyclicBarrier barrier;
+      final RemoteCache<Integer, String> remote;
+      final int threadId;
```
Comment (wburns) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java]: Is threadId and testName for debugging purposes?  Seems they are never used.

File: commons/src/main/java/org/infinispan/commons/io/UnsignedNumeric.java [hunk 5]
```diff
@@ -170,12 +170,14 @@ public static int readUnsignedInt(byte[] bytes, int offset) {
     *
     * @param i int to write
     */
-   public static void writeUnsignedInt(byte[] bytes, int offset, int i) {
+   public static int writeUnsignedInt(byte[] bytes, int offset, int i) {
```
Comment (wburns) [inline commons/src/main/java/org/infinispan/commons/io/UnsignedNumeric.java]: :+1: I am surprised we didn't have this before

I wonder if returning the size used may be better, WDYT?

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala [hunk 6]
```diff
@@ -23,19 +22,39 @@ object KeyValueVersionConverterFactory {
...
+   private def putBytes(bytes: Array[Byte], offset: Int, out: Array[Byte]): Int = {
+      var localOffset = offset
+      bytes.foreach(b => {
+         out.update(localOffset, b)
+         localOffset += 1
+      })
+      localOffset
+   }
+
+   private def putLong(l: Long, offset: Int, out: Array[Byte]): Int = {
```
Comment (wburns) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala]: Can the version be negative?  Wondering why we didn't use UnsignedNumeric.writeUnsignedLong.  Granted we can't change it now since it is part of the protocol spec.

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala (hunk 1 above)
Comment (gustavonalle) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala]: in scala the value of the last statement in the method body is returned

Comment (wburns) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala]: Ah yes sorry I missed the Int return definition on line 46 :)

**Comment by galderz** [comment]
I'm separating the "Use MessageToByteEncoder to avoid ByteBuf leak w/ exceptions" part in a different JIRA + PR, since it's been spotted by a user.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java (hunk 4 above)
Comment (galderz) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java]: Yeah, for debugging. Not used now, so I'll remove.

File: commons/src/main/java/org/infinispan/commons/io/UnsignedNumeric.java (hunk 5 above)
Comment (galderz) [inline commons/src/main/java/org/infinispan/commons/io/UnsignedNumeric.java]: > I wonder if returning the size used may be better, WDYT?

I've tried that but makes the client usage more complex since you need to keep both written bytes and offset separately. Offset is needed to figure out where to start writing in the byte[]. Returning the new offset makes it simpler IMO.

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala (hunk 6 above)
Comment (galderz) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala]: Could be negative in a cluster environment since it's built of different parts. The version is a 64 bit long inspired originally by Memcached, and we decided to keep that way since we'd likely to use the 64 bits in a cluster env (see `org.infinispan.container.versioning.NumericVersionGenerator`). Only in local envs the number will be small enough for a time to make `writeUnsignedLong` worth it.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java (hunk 3 above)
Comment (galderz) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java]: Removed

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java (hunk 2 above)
Comment (galderz) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/stress/near/EagerNearCacheStressTest.java]: Done

**Comment by galderz** [comment]
Updated, I added the small fixes to the last commit by mistake but it should be fine.

**Comment by danberindei** [comment]
@galderz I'm seeing some failures in EagerNearCacheTest on my machine, e.g.:

    testGetNearCacheAfterConnect(org.infinispan.client.hotrod.near.EagerNearCacheTest)  Time elapsed: 0.54 sec  <<< FAILURE!
    java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertNotNull(AssertJUnit.java:267)
	at org.testng.AssertJUnit.assertNotNull(AssertJUnit.java:259)
	at org.infinispan.client.hotrod.near.AssertsNearCache.pollEvent(AssertsNearCache.java:153)
	at org.infinispan.client.hotrod.near.AssertsNearCache.expectNearRemoveInClient(AssertsNearCache.java:145)
	at org.infinispan.client.hotrod.near.AssertsNearCache.expectNearRemove(AssertsNearCache.java:119)
	at org.infinispan.client.hotrod.near.EagerNearCacheTest.testGetNearCacheAfterConnect(EagerNearCacheTest.java:53)

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala [hunk 7]
```diff
@@ -23,19 +22,39 @@ object KeyValueVersionConverterFactory {
          val capacity = UnsignedNumeric.sizeUnsignedInt(key.length) + key.length +
             (if (newValue != null) UnsignedNumeric.sizeUnsignedInt(newValue.length) + newValue.length + 8 else 0)
 
-         val out = ByteBuffer.allocate(capacity)
-         UnsignedNumeric.writeUnsignedInt(out, key.length)
-         out.put(key)
+         val out = Array.ofDim[Byte](capacity)
```
Comment (danberindei) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala]: Is this change really necessary? The overhead of ByteBuffer seems minimal to me, and o.i.commons.io.ByteBuffer has even less fields...

**Comment by danberindei** [comment]
Eager near cache tests failing in CI as well: http://ci.infinispan.org/viewLog.html?buildId=27977&buildTypeId=bt9&guest=1

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala (hunk 7 above)
Comment (galderz) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala]: Just trying to do the least amount of possible work here really. There's little to be built up really, so the least resources consumed for it, the better.

**Comment by galderz** [comment]
I'm looking into the test failures

**Comment by galderz** [comment]
@danberindei Fixed the test failures

**Comment by wburns** [comment]
Pulling..

**Comment by wburns** [comment]
Integrated into master, thanks @galderz !
